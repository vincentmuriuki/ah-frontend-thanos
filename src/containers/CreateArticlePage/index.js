import React, { Component } from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { postArticleData, createArticleThunk, updateImageUrl } from '../../actions/createArticleActions';
import CreateArticle from '../../components/CreateArticle';


class CreateArticlePage extends Component {
  handleChange = (event) => {
    event.preventDefault();
    const { articleData, createArticleReducer } = this.props;
    articleData({
      ...createArticleReducer.articlePostData,
      [event.target.name]: event.target.value,
    });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { postArticle, createArticleReducer } = this.props;
    postArticle({ ...createArticleReducer.articlePostData, tag_list: createArticleReducer.articlePostData.tag_list.split(', ') });
  }

  handleOnClick = () => {
    window.cloudinary.openUploadWidget({
      cloud_name: 'author-s-haven',
      upload_preset: 'cwqbp3ae',
      cropping: true,
      folder: 'widgetdocs',
      sources: ['local', 'url', 'camera', 'facebook', 'dropbox', 'search', 'instagram'],
    }, (error, result) => {
      if (result.event === 'success') {
        const { updateImage } = this.props;
        updateImage(result.info.secure_url);
      }
    });
  }

  render() {
    return (
      <div>
        <CreateArticle
          onChange={this.handleChange}
          onSubmit={this.handleSubmit}
          onClick={this.handleOnClick}
        />
      </div>
    );
  }
}

CreateArticlePage.propTypes = {
  postArticle: PropTypes.func.isRequired,
  articleData: PropTypes.func.isRequired,
  createArticleReducer: PropTypes.shape({}).isRequired,
  updateImage: PropTypes.func.isRequired,
};


const mapStateToProps = ({ createArticleReducer }) => ({ createArticleReducer });

export const mapDispatchToProps = dispatch => ({
  postArticle: response => dispatch(createArticleThunk(response)),
  articleData: response => dispatch(postArticleData(response)),
  updateImage: response => dispatch(updateImageUrl(response)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateArticlePage);
