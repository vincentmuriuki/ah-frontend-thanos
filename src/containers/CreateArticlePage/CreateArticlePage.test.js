import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CreateArticlePage, { mapDispatchToProps } from './index';

describe('<CreateArticlePage />', () => {
  let CreateArticlePageComponent;
  let wrapper;
  beforeEach(() => {
    jest.resetModules();
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    const initialState = { createArticleReducer: { articlePostData: { tag_list: 'one,two', title: 1 } } };
    const store = mockStore(initialState);
    CreateArticlePageComponent = mount(
      <CreateArticlePage store={store} />,
    );
    wrapper = shallow(
      <CreateArticlePage store={store} />,
    );
  });

  test('test function ', async () => {
    global.cloudinary = {
      openUploadWidget: (params, cb) => {
        cb(null, {
          event: 'success',
          info: { secure_url: 'http://cloudinary/img/123.png' },
        });
      },
    };
    const widgetFn = wrapper.dive().instance().handleOnClick();
    expect(widgetFn);
  });

  it('should render the component', () => {
    expect(CreateArticlePageComponent).toMatchSnapshot();
  });
  const dispatchTest = dispatch => (
    mapDispatchToProps(dispatch).postArticle({})
    && mapDispatchToProps(dispatch).articleData({})
    && mapDispatchToProps(dispatch).updateImage({})
  );
  it('should dispatch a method to get user input', () => {
    const dispatch = jest.fn();
    dispatchTest(dispatch);
  });

  it('should test user data without fail', () => {
    CreateArticlePageComponent
      .find('textarea#titleInput')
      .simulate('change', { target: { name: 'title', value: 'How ro go to school' } });
    CreateArticlePageComponent
      .find('form')
      .simulate('submit');
    CreateArticlePageComponent
      .find('input.image')
      .simulate('change', { target: { name: 'image_url', value: 'http://image.com' } });
  });
});
