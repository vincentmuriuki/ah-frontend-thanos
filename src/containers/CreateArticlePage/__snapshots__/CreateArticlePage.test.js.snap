// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateArticlePage /> should render the component 1`] = `
<Connect(CreateArticlePage)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CreateArticlePage
    articleData={[Function]}
    createArticleReducer={
      Object {
        "articlePostData": Object {
          "tag_list": "one,two",
          "title": 1,
        },
      }
    }
    postArticle={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    updateImage={[Function]}
  >
    <div>
      <CreateArticle
        onChange={[Function]}
        onClick={[Function]}
        onSubmit={[Function]}
      >
        <div>
          <br />
          <h3
            className="text-center createArticleTitle"
          >
            Create Your Article
          </h3>
          <hr
            className="col-sm-8 "
          />
          <br />
          <div
            className="col-sm-10 offset-2 artcileCreate"
          >
            <form
              className="create-article-form"
              onSubmit={[Function]}
            >
              <h5>
                Title: 
              </h5>
              <small>
                Title must be 130 characters maximum and 5 characters minimum
              </small>
              <div
                className="form-group"
              >
                <textarea
                  className="form-control"
                  id="titleInput"
                  maxLength="130"
                  minLength="5"
                  name="title"
                  onChange={[Function]}
                  required={true}
                  rows="3"
                />
              </div>
              <h5>
                Description
              </h5>
              <small>
                Title must be 150 characters maximum and 100 characters minimum
              </small>
              <div
                className="form-group"
              >
                <textarea
                  className="form-control"
                  id="descriptionInput"
                  maxLength="120"
                  minLength="100"
                  name="description"
                  onChange={[Function]}
                  required={true}
                  rows="3"
                />
              </div>
              <h5>
                Tag list
              </h5>
              <div
                className="form-group"
              >
                <textarea
                  className="form-control"
                  id="tag_listInput"
                  name="tag_list"
                  onChange={[Function]}
                  required={true}
                  rows="3"
                />
              </div>
              <br />
              <h5>
                Upload Image
              </h5>
              <div
                className="col-lg-8 mt-2"
              >
                <input
                  className="image"
                  name="image_url"
                  onClick={[Function]}
                  type="file"
                />
              </div>
              <br />
              <h5>
                Body
              </h5>
              <div
                className="form-group"
              >
                <textarea
                  className="form-control"
                  id="bodyInput"
                  name="body"
                  onChange={[Function]}
                  required={true}
                  rows="3"
                />
              </div>
              <div
                className="signin-createArticleBt"
              >
                <button
                  className="btn btn-primary"
                  id="createArticleBt"
                  name="createArticleBt"
                  type="submit"
                >
                  Publish Article
                </button>
              </div>
            </form>
          </div>
        </div>
      </CreateArticle>
    </div>
  </CreateArticlePage>
</Connect(CreateArticlePage)>
`;
